src/systems/joint_trajectory_controller/JointTrajectoryController.cc

```diff
@@ -0,0 +1,792 @@
+        {
+          // Warning about unconfigured joint is already logged above
+          continue;
+        }
+        auto *joint = &this->dataPtr->actuatedJoints[jointName];
+        joint->SetTarget(targetPoint, jointIndex);
+
+        // Reset also the PID error of the affected joints
+        joint->ResetPIDs();
+      }
+
+      // If there are no more points after the current one, set the trajectory to Reached
+      if (this->dataPtr->trajectory.IsGoalReached())
+      {
+        this->dataPtr->trajectory.status = Trajectory::Reached;
+      }
+
+      // Publish current progress of the trajectory
+      ignition::msgs::Float progressMsg;
+      progressMsg.set_data(this->dataPtr->trajectory.ComputeProgress());
+      this->dataPtr->progressPub.Publish(progressMsg);
+    }
+  }
+
+  // Control loop
+  for (auto &actuatedJoint : this->dataPtr->actuatedJoints)
+  {
+    auto *joint = &actuatedJoint.second;
+    joint->Update(_ecm, _info.dt);
+  }
+}
+
+////////////////////////////////////////
+/// JointTrajectoryControllerPrivate ///
+////////////////////////////////////////
+
+void JointTrajectoryControllerPrivate::ConfigureJoint(
+    const Entity &_entity,
+    const std::shared_ptr<const sdf::Element> &_sdf,
+    const ignition::gazebo::EntityComponentManager &_ecm,
+    const std::vector<std::string> &_enabledJoints)
+{
+  const auto jointName = _ecm.Component<components::Name>(_entity)->Data();
+  auto jointIndex = this->actuatedJoints.size();
+
+  // Ignore duplicate joints
+  for (const auto &actuatedJoint : this->actuatedJoints)
+  {
+    if (actuatedJoint.second.entity == _entity)
+    {
+      ignwarn << "[JointTrajectoryController] Ignoring duplicate joint [" << jointName << "(Entity="
+              << _entity << ")]\".\n";
+      continue;
+    }
+  }
+
+  // Make sure the joint type is supported, i.e. it has a single actuated axis
+  const auto *jointType = _ecm.Component<components::JointType>(_entity);
+  switch (jointType->Data())
+  {
+  case sdf::JointType::PRISMATIC:
+  case sdf::JointType::REVOLUTE:
+  case sdf::JointType::CONTINUOUS:
+  case sdf::JointType::GEARBOX:
+  {
+    // Supported joint type
+    break;
+  }
+  case sdf::JointType::FIXED:
+  {
+    igndbg << "[JointTrajectoryController] Fixed joint [" << jointName << "(Entity=" << _entity
+           << ")] is skipped.\n";
+    return;
+  }
+  case sdf::JointType::REVOLUTE2:
+  case sdf::JointType::SCREW:
+  case sdf::JointType::BALL:
+  case sdf::JointType::UNIVERSAL:
+  {
+    ignwarn << "[JointTrajectoryController] Joint [" << jointName << "(Entity=" << _entity
+            << ")] is of unsupported type."
+            << " Only joints with a single axis are supported.\n";
+    return;
+  }
+  default:
+  {
+    ignwarn << "[JointTrajectoryController] Joint [" << jointName << "(Entity=" << _entity
+            << ")] is of unknown type.\n";
+    return;
+  }
+  }
+
+  // Skip if joint is not enabled
```

Comment:
Should we check this before checking the joint types? Imagine the model has a revolute joint and a ball joint. The user knows they can't control the ball joint, so they don't enable it. But they see the warning anyway, right?

Reply:
Very true, I refactored this part in https://github.com/ignitionrobotics/ign-gazebo/pull/473/commits/54c8d7da607f7ff4237b1b119ce04f8aaf2a0d75 to accommodate new param parsing and allow arbitrary order for joint names. It should now also give an error if model does not contain a requested joint name.